src/dune_engine/cached_digest.ml

```diff
@@ -332,7 +332,7 @@ let remove path =
   Path.Table.remove cache.table path
 
 module Untracked = struct
-  let source_or_external_file = peek_or_refresh_file ~allow_dirs:false
+  let source_or_external_path = peek_or_refresh_file ~allow_dirs:true
```

Comment:
I think the call site should be aware about whether the path is a file or a directory, so it's safer and clearer for reding the call sites to have separate functions: one for files, one for directories. I understand that loose typing is convenient but it also comes at a cost of readability/correctness.

Reply:
I say "loose typing" above, which I guess is not fair because we don't distinguish files from directories at the moment. But we should -- that would catch a ton of errors! And once we do distinguish them, we'll need to split this function in two anyways, which is why I suggest doing this right now.

Reply:
What about adding a `kind = Directory | File` argument to all the relevant functions? It should make things as explicit while not expanding the interface by 2x.

Reply:
I think I prefer `file_`, `dir_` and `path_` because `Fs_memo.file_exists` reads better than `Fs_memo.exists ~kind:File`. I suspect it "writes better" too. 

It's not like our `x` in `2x` is very large: it's like 2 or maybe 3 at a stretch? In some cases `dir_` functions have no `file_` counterpart, so we'd end up with some inconsistent naming anyway.

Reply:
> In some cases dir_ functions have no file_ counterpart, so we'd end up with some inconsistent API anyway.

Presumably, once we have `Path.Directory.t`, the types would reflect this anyway.

Reply:
Marking this as resolved since this part of the PR has been dropped.